Нечуговских Антон

TonyCardio opened the pull request:
@lurlrlrl

TonyCardio committed:
add TagsCloudVisualization proj

TonyCardio committed:
complete "main" task

TonyCardio committed:
complete "Aesthetic" task

TonyCardio committed:
complete "research" task

lurlrlrl commented on cs/TagsCloudVisualization/Extensions/PointFComparer.cs:
```diff
@@ -0,0 +1,20 @@
+
+        public int GetHashCode(PointF point)
+        {
+            return point.GetHashCode();
```
Текущая реализация опасна тем, что у двух разных точек может быть разный хэш код, но эти точки могут быть равны

lurlrlrl commented on cs/TagsCloudVisualization/Tests/ArchimedeanSpiralTests.cs:
```diff
@@ -0,0 +1,43 @@
+        [Test]
+        [TestCase(2, TestName = "GetTwoPoints")]
+        [TestCase(10, TestName = "GetTenPoints")]
+        [TestCase(50, TestName = "GetFiftyPoints")]
```
Не увидел смысла в явном задании TestName в текущем случае, по-моему и без него понятно

lurlrlrl commented on cs/TagsCloudVisualization/Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,128 @@
+        }
+
+        [Test]
+        [TestCase(0, 0, TestName = "CenterInStartOfCoordinates")]
```
Я бы лучше использовал слово origin, а не start

lurlrlrl commented on cs/TagsCloudVisualization/Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,128 @@
+            layouter = new CircularCloudLayouter(center);
+        }
+
+        [Test]
```
Не нужно писать Test, когда есть TestCase

lurlrlrl commented on cs/TagsCloudVisualization/Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,128 @@
+            Console.WriteLine($"Tag cloud visualization saved to file {filePath}");
+        }
+
+        private bool HaveAnyIntersections(List<Rectangle> rects)
```
Метод можно сделать статичным

lurlrlrl commented on cs/TagsCloudVisualization/Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,128 @@
+
+        private bool HaveAnyIntersections(List<Rectangle> rects)
+        {
+            for (int i = 0; i < rects.Count; i++)
```
Можно написать с помощью foreach или LINQ

TonyCardio committed:
refactor

* (Extentions, Models, Tests, View)
    minor changes
* (Readme,Images)
    update image in readme

tests

* (CircularCloudLayouterTests)
    add placement density test

lurlrlrl commented on cs/TagsCloudVisualization/Models/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,84 @@
+                throw new ArgumentException("Layout is empty");
+
+            var deltaX = rectangles
+                .Max(rectangle => rectangle.Right) - center.X;
```
А почему здесь считается самая правая точка, но не считается самая левая?

lurlrlrl commented on cs/TagsCloudVisualization/Models/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,84 @@
+using System.Drawing;
+using System.Linq;
+
+namespace TagsCloudVisualization.Models
```
А почему папка называется Models?

lurlrlrl commented on cs/TagsCloudVisualization/View/TagCloudCreator.cs:
```diff
@@ -0,0 +1,54 @@
+            var filePath = Path.Combine(dirPath, filename);
+            var bitmap = GetFilledBitmap();
+            bitmap.Save(filePath, ImageFormat.Png);
+            bitmap.Dispose();
```
А если выпало иссключение? Нужно использовать using

lurlrlrl commented on cs/TagsCloudVisualization/View/TagCloudCreator.cs:
```diff
@@ -0,0 +1,54 @@
+        {
+            var layoutSize = layouter.GetLayoutSize();
+            var bitmap = new Bitmap(layoutSize.Width, layoutSize.Height);
+            var graphics = Graphics.FromImage(bitmap);
```
Graphics это тоже IDisposable

lurlrlrl commented on cs/TagsCloudVisualization/Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,141 @@
+            int rectsCount)
+        {
+            PutRectangles(rectsCount);
+            layouter.Rectangles.ToList().Count.Should().Be(rectsCount);
```
Можно просто вызвать метод Count(), без ToList()

lurlrlrl commented on cs/TagsCloudVisualization/Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,141 @@
+            int width = 10, int height = 10)
+        {
+            for (int i = 0; i < amountToPlace; i++)
+                layouter.PutNextRectangle(new Size(width, height));
```
Метод PutNextRectangle возвращает Rectangle, я бы сразу сохранял все прямоугольники в поле, вместо вызова в тесте layouter.Rectangles.ToList()

lurlrlrl commented on cs/TagsCloudVisualization/Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,141 @@
+
+            var expectedRadius = Math.Sqrt(layoutArea / Math.PI);
+            var deltaXFromCenter = (double)layouter.Rectangles
+                .Max(rectangle => rectangle.Right) - center.X;
```
А почему здесь считается самая правая точка, но не считается самая левая?

TonyCardio committed:
refactor

* (TagCloudCreator,CircularCloudLayouterTests)
    minor changes

tests

* (CircularCloudLayouterTests)
    fix cloud shape test

lurlrlrl commented on cs/TagsCloudVisualization/View/TagCloudCreator.cs:
```diff
@@ -0,0 +1,59 @@
+                foreach (var rectangle in layouter.Rectangles)
+                    graphics.DrawRectangle(rectanglePen, rectangle);
+
+                layoutBrush.Dispose();
```
А почему метод Save можно вызвать несколько раз, а layoutBrush и rectanglePen диспоузится после первого вызова?
Нужно сделать либо что-бы они создавались и диспоузились на каждый вызов .Save, либо сделать TagCloudCreator IDisposable и диспоузить их вместе с ним

TonyCardio committed:
fixes

* (TagCloudCreator)
    fix bad using IDisposable objects

refactor

* (Program)
    minor changes

TonyCardio committed:
feature

* (Extensions)
    add extension to find distance to layout border

fixes

* (TagCloudCreator)
    fix problem with brush disposing

refactor

* (Models, Tests)
    minor changes

lurlrlrl commented on cs/TagsCloudVisualization/View/TagCloudCreator.cs:
```diff
@@ -48,12 +48,15 @@ private Bitmap GetFilledBitmap()
+
+        public void Dispose()
+        {
+            layoutBrush?.Dispose();
```
А `?` здесь не лишний?

TonyCardio committed:
refactor

* (TagCloudCreator)
    delete nullable links
